File: backend/app/schemas/analytics.py
"""
Analytics schemas for request/response validation.
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel


class AnalyticsEventCreate(BaseModel):
    """Analytics event creation schema."""
    event_type: str
    event_category: str
    resource_id: Optional[int] = None
    extra_data: Optional[Dict[str, Any]] = None


class AnalyticsEventResponse(BaseModel):
    """Analytics event response schema."""
    id: int
    user_id: Optional[int] = None
    event_type: str
    event_category: str
    resource_id: Optional[int] = None
    extra_data: Optional[Dict[str, Any]] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ActivityStats(BaseModel):
    """Overall activity statistics."""
    total_events: int
    events_today: int
    events_this_week: int
    events_this_month: int
    unique_users: int


class PopularCode(BaseModel):
    """Popular diagnostic code statistics."""
    code_id: int
    code: str
    description: str
    view_count: int


class EventTypeDistribution(BaseModel):
    """Distribution of event types."""
    event_type: str
    count: int


class UserActivity(BaseModel):
    """User activity statistics."""
    user_id: int
    username: str
    email: str
    event_count: int
    last_activity: datetime


class AnalyticsSummary(BaseModel):
    """Complete analytics summary."""
    activity_stats: ActivityStats
    popular_codes: List[PopularCode]
    event_distribution: List[EventTypeDistribution]
    recent_events: List[AnalyticsEventResponse]
    top_users: List[UserActivity]

File: frontend/src/services/analytics.ts
import apiClient from '@/lib/apiClient';

export interface AnalyticsEventCreate {
  event_type: string;
  event_category: string;
  resource_id?: number;
  metadata?: Record<string, any>;
}

export interface ActivityStats {
  total_events: number;
  events_today: number;
  events_this_week: number;
  events_this_month: number;
  unique_users: number;
}

export interface PopularCode {
  code_id: number;
  code: string;
  description: string;
  view_count: number;
}

export interface EventTypeDistribution {
  event_type: string;
  count: number;
}

export interface UserActivity {
  user_id: number;
  username: string;
  email: string;
  event_count: number;
  last_activity: string;
}

export interface AnalyticsSummary {
  activity_stats: ActivityStats;
  popular_codes: PopularCode[];
  event_distribution: EventTypeDistribution[];
  recent_events: any[];
  top_users: UserActivity[];
}

class AnalyticsService {
  async trackEvent(event: AnalyticsEventCreate): Promise<void> {
    try {
      await apiClient.post('/api/v1/analytics/events', event);
    } catch (error) {
      console.error('Failed to track analytics event:', error);
    }
  }

  async getActivityStats(): Promise<ActivityStats> {
    const response = await apiClient.get<ActivityStats>('/api/v1/analytics/activity');
    return response.data;
  }

  async getPopularCodes(limit: number = 10): Promise<PopularCode[]> {
    const response = await apiClient.get<PopularCode[]>('/api/v1/analytics/popular-codes', {
      params: { limit }
    });
    return response.data;
  }

  async getEventDistribution(): Promise<EventTypeDistribution[]> {
    const response = await apiClient.get<EventTypeDistribution[]>('/api/v1/analytics/event-distribution');
    return response.data;
  }

  async getAnalyticsSummary(): Promise<AnalyticsSummary> {
    const response = await apiClient.get<AnalyticsSummary>('/api/v1/analytics/summary');
    return response.data;
  }

  async getTopUsers(limit: number = 10): Promise<UserActivity[]> {
    const response = await apiClient.get<UserActivity[]>('/api/v1/analytics/top-users', {
      params: { limit }
    });
    return response.data;
  }
}

export const analyticsService = new AnalyticsService();
